cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/AbstractInMemoryProgramRunner.java

```diff
@@ -144,14 +145,33 @@ private void monitorComponents() {
         controller.addListener(new AbstractListener() {
           @Override
           public void completed() {
-            if (liveComponents.decrementAndGet()  == 0) {
-              complete();
+            if (liveComponents.decrementAndGet() == 0) {
+              // If at least one of the components failed with an error, then set the state to ERROR
+              if (errorCause != null) {
+                setError(errorCause);
+              } else {
+                complete();
+              }
+            }
+          }
+
+          @Override
+          public void error(Throwable cause) {
```

Comment:
What if the instance is stopped? It should get marked as `KILLED`, right? So we need to handle that case as well.